File: examples/Actor.Next/06-Approvals/Approvals.Next.Example06.Tests/VerificationTests.cs
using Dapr.Actors.Next.Abstractions;
using Dapr.Actors.Next.Interpreted;
using Dapr.Workflow;
using Moq;

namespace Dapr.Actors.Next.Examples.Approvals.Tests;

public sealed class VerificationTests
{
    private static readonly ActorId Expense = ActorId.Create("exp-1");

    [Fact]
    public void Well_formed_definition_passes_verification()
    {
        var verifier = new InterpretedMachineVerifier(Registry());

        Assert.True(ApprovalDefinitions.Verify(verifier, ApprovalDefinitions.ExpenseReport()).IsValid);
        Assert.True(ApprovalDefinitions.Verify(verifier, ApprovalDefinitions.Contract()).IsValid);
    }

    [Fact]
    public void Verification_rejects_definition_that_can_strand_an_approved_document()
    {
        var verifier = new InterpretedMachineVerifier(Registry());
        var stranded = ApprovalDefinitions.ExpenseReport(includeSettlementCompletion: false);

        var result = ApprovalDefinitions.Verify(verifier, stranded);

        Assert.False(result.IsValid);
        Assert.Contains(result.Defects, defect => defect.Contains("State 'Approved' is a dead end", StringComparison.Ordinal));
    }

    [Fact]
    public async Task Definition_referencing_an_unregistered_effect_is_rejected_before_rollout()
    {
        var registry = Registry();
        var deployer = new InterpretedMachineDeployer(new InterpretedMachineVerifier(registry), new InMemoryInterpretedMachineStore());
        var definition = ApprovalDefinitions.ExpenseReport(settlementEffect: "UnknownSettlement");

        Assert.False(registry.TryGetEffect("UnknownSettlement", out _));
        var ex = await Assert.ThrowsAsync<InvalidOperationException>(() =>
            deployer.DeployAsync(ApprovalDefinitions.ActorType, Expense, definition).AsTask());
        Assert.Contains("Effect 'UnknownSettlement'", ex.Message, StringComparison.Ordinal);
    }

    // The verifier only needs the registry to resolve capability names; the workflow client is never called.
    private static ApprovalCapabilityRegistry Registry() =>
        new(Mock.Of<IDaprWorkflowClient>());
}

File: examples/Actor.Next.FSharp/06-Approvals/Approvals.Next.FSharp.Example06.Tests/VerificationTests.fs
namespace Approvals.Next.FSharp.Example06.Tests

open System
open System.Threading.Tasks
open Dapr.Actors.Next.Abstractions
open Dapr.Actors.Next.Abstractions.Filters
open Dapr.Actors.Next.Interpreted
open Dapr.Actors.Next.Examples.Approvals
open Dapr.Workflow
open Xunit

type VerificationTests() =

    static member private Expense = ActorId.Create("exp-1")

    member private _.Registry() : ApprovalCapabilityRegistry =
        ApprovalCapabilityRegistry(new RecordingWorkflowClient())

    [<Fact>]
    member this.Well_formed_definition_passes_verification() =
        let verifier = InterpretedMachineVerifier(this.Registry() :> ICapabilityRegistry)
        Assert.True(ApprovalDefinitions.Verify(verifier, ApprovalDefinitions.ExpenseReport()).IsValid)
        Assert.True(ApprovalDefinitions.Verify(verifier, ApprovalDefinitions.Contract()).IsValid)

    [<Fact>]
    member this.Verification_rejects_definition_that_can_strand_an_approved_document() =
        let verifier = InterpretedMachineVerifier(this.Registry() :> ICapabilityRegistry)
        let stranded = ApprovalDefinitions.ExpenseReport(includeSettlementCompletion = false)

        let result = ApprovalDefinitions.Verify(verifier, stranded)

        Assert.False(result.IsValid)
        let containsDeadEnd =
            result.Defects
            |> Seq.exists (fun d -> d.Contains("State 'Approved' is a dead end", StringComparison.Ordinal))
        Assert.True(containsDeadEnd)

    [<Fact>]
    member this.Definition_referencing_an_unregistered_effect_is_rejected_before_rollout() = task {
        let registry = this.Registry()
        let deployer = InterpretedMachineDeployer(
            InterpretedMachineVerifier(registry :> ICapabilityRegistry) :> IInterpretedMachineVerifier,
            InMemoryInterpretedMachineStore() :> IInterpretedMachineStore)
        let definition = ApprovalDefinitions.ExpenseReport(settlementEffect = "UnknownSettlement")

        let capRegistry = registry :> ICapabilityRegistry
        let mutable effect = Unchecked.defaultof<IActorEffect>
        Assert.False(capRegistry.TryGetEffect("UnknownSettlement", &effect))

        let! ex = Assert.ThrowsAsync<InvalidOperationException>(fun () ->
            deployer.DeployAsync(ApprovalDefinitions.ActorType, VerificationTests.Expense, definition).AsTask())
        Assert.Contains("Effect 'UnknownSettlement'", ex.Message, StringComparison.Ordinal)
    }
